### PR TITLE
[Layout] Wiki nav with custom items and TOC tree

### PR DESCRIPTION
**Pull request by wielinde**
Still left todo

- [ ] Mark current menu item as selected

### CODE REVIEW

File: app/assets/javascripts/specific/main_menu.js.erb [hunk 1]
```diff
@@ -63,23 +46,18 @@ jQuery(document).ready(function($) {
...
-        if (menuItemWrapper.hasClass('open')) {
-          menuParent.show();
-          menuParent.find('li > a:first').focus();
-        } else {
-          menuParent.hide();
-        }
-
-        setExpandStatus(target);
+
+        // TODO: Instead of hiding the sidebar move sidebar's contents to submenus and cache it.
```
Comment (HDinger) [inline app/assets/javascripts/specific/main_menu.js.erb:54]: Are these TODOs out of scope for this PR?

File: app/assets/javascripts/tree_menu.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// See docs/COPYRIGHT.rdoc for more details.
+//++
+
+(function ($) {
+  "use strict";
+  $(function() {
+    $('.pages-hierarchy.-with-hierarchy .tree-menu--hierarchy-span').click(function (event) {
+      var target = $(event.target);
+      var targetList = target.closest('.-with-hierarchy');
+      console.log("click was triggered", targetList);
```
Comment (HDinger) [inline app/assets/javascripts/tree_menu.js]: Remove the `console.log`

File: app/assets/javascripts/tree_menu.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      if (targetList.hasClass('-hierarchy-collapsed')) {
```
Comment (HDinger) [inline app/assets/javascripts/tree_menu.js]: I guess more elegant would be to work with `toggle` (`targetList.toggle('-hierarchy-collapsed');`, `targetList.toggle('-hierarchy-expanded');`.
Thus you can avoid the `if`..

File: app/assets/stylesheets/layout/_main_menu_mobile.sass [hunk 4]
```diff
@@ -27,22 +27,23 @@
 //++
 
 @include breakpoint(680px down)
-  #toggle-project-menu.show + #menu-sidebar
-    display: none
 
-  #main-menu
+  .main-menu
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_main_menu_mobile.sass:31]: Why did you change it to a class? Can there be more than one main-menu?

File: app/assets/stylesheets/layout/_top_menu.sass [hunk 5]
```diff
@@ -100,7 +97,7 @@
         float: none
     > ul.drop-down--projects
       width: 400px
-      left: 0
+      left: -2px
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_top_menu.sass:103]: Why is this necessary? Is there a way to avoid it?

**Review by HDinger — CHANGES_REQUESTED** [review]

**Comment by HDinger** [comment]
For me the resizing does not work on MS Edge. Collapse and Expand via click works fine, but not drag'n'drop.

**Comment by HDinger** [comment]
There are some accessibility issues left, I want to mention so that we do not forget about them:

* The label to expand/collapse of the hamburger and the resizer icon should change according to the state.
* The arrows to collapse a hierarchy tree cannot be used via keyboard.
* After tabing through the details navigation (of the WP) the focus gets lost. (When the details are finished the content should be focused)

I will create a ticket and assigned these issues to @ingamai

**Comment by HDinger** [comment]
When the sidebar is collapsed, the elements can still be reached via keyboard. Please explicitly hide them so that cannot happen (this is also an accessibility issue btw).

**Comment by HDinger** [comment]
@wielinde To be honest I did not have the time to dig very deep into the code changes. It made however a good impression. If you are not sure and do not want to wait until next Friday you should ask someone else to have a look at it, too.

**Comment by HDinger** [comment]
Please have in mind that some plugins also have detailed navigation (e.g. Gitolite)

File: app/assets/stylesheets/layout/_top_menu.sass (hunk 5 above)
Comment (wielinde) [reply app/assets/stylesheets/layout/_top_menu.sass:103]: I thought it was prettier  to have the 2px grey border of the drop down menu align be "outside" so that the white of the drop-down aligns with the hover background of the top menu item. I felt that the border has more of a shadow and thus belongs outside. Remember that before there was no hamburger icon and the issue was not so apparent as the border was directly hitting the viewport edge. Anyways, if you guys think it does not improve anything feel free to kill it.

File: app/assets/stylesheets/layout/_main_menu_mobile.sass (hunk 4 above)
Comment (wielinde) [reply app/assets/stylesheets/layout/_main_menu_mobile.sass:31]: The tree-menu "component" is generic and has its own styles. However, every CSS rule below an ID selector has precedence which means that I'd have to repeat every style for the tree component to overwrite other CSS rules the were set below the ID selector.

I think it is ok to use ID selectors on very specific CSS rules. But it makes things complicated when there are tons of rules cascading below the ID identifier. I think using ID selectors is great for jQuery stuff but it should be handled with care when it comes to CSS.

File: app/assets/javascripts/specific/main_menu.js.erb (hunk 1 above)
Comment (wielinde) [reply app/assets/javascripts/specific/main_menu.js.erb:54]: Yes, out of scope, but wanted.

**Comment by wielinde** [comment]
@HDinger I cannot reproduce Resizing issues on MS Edge (42.17134.1.0). What version are you on?

**Comment by oliverguenther** [comment]
Feels very solid :+1: I have only one additional minor remark apart from the code review.

Long project names now can now leak into the OpenProject logo, you may want to reduce its maximum width a little again.

screenshot from 2018-05-23 09-46-53


**Accessibility** (we may want to move this into a separate PR)
- [ ] Both the hamburger div and link can be tabbed into, this should be avoided
- [ ] As Henriette wrote, the menus are still accessible by tab even when collapsed.
- [ ] As Heneriette wrote, The arrows to collapse a hierarchy tree can be reached, but not activated via keyboard
- [x] After tabbing through the details navigation (of the WP) the focus gets lost. (When the details are finished the content should be focused) *seems to be fixed*

File: app/assets/javascripts/tree_menu.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      if (selectedPage) {
+        $('[slug=' + selectedPage + ']', tree).toggleClass('-selected', true);
+      }
+    });
+
+    // set click handlers for expanding and collapsing tree nodes
+    $('.pages-hierarchy.-with-hierarchy .tree-menu--hierarchy-span').click(function (event) {
+      var target = $(event.target);
+      var targetList = target.closest('.-with-hierarchy');
+      if (targetList.hasClass('-hierarchy-collapsed')) {
```
Comment (oliverguenther) [inline app/assets/javascripts/tree_menu.js]: This whole condition until line 50 should be the same as ` targetList.toggleClass('-hierarchy-collapsed -hierarchy-expanded')`

File: app/assets/stylesheets/layout/_main_menu.sass [hunk 7]
```diff
@@ -57,124 +55,34 @@ $toggler-width: 40px
...
-
-              &:hover
-                @include varprop(background, main-menu-bg-hover-background)
-
-            &.selected, &.selected + a
-              @include varprop(color, main-menu-selected-font-color)
-
-            &:hover
-              @include varprop(background, main-menu-bg-hover-background)
-              @include varprop(color, main-menu-hover-font-color)
```
Comment (oliverguenther) [inline app/assets/stylesheets/layout/_main_menu.sass:109]: :muscle:

File: app/assets/stylesheets/layout/_main_menu_mobile.sass (hunk 4 above)
Comment (oliverguenther) [reply app/assets/stylesheets/layout/_main_menu_mobile.sass:31]: We should never use IDs for CSS rules, so this is absolutely correct

File: app/assets/stylesheets/layout/_top_menu.sass (hunk 5 above)
Comment (oliverguenther) [reply app/assets/stylesheets/layout/_top_menu.sass:103]: When switching between the -2px and 0px, I too prefer the former

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -573,6 +570,59 @@ def permitted_params
...
+  def hierarchy_item_title(options, page)
+    if options[:timestamp] && page.updated_on
+      l(:label_updated_time, distance_of_time_in_words(Time.now, page.updated_on))
+    end
+  end
+
+  def render_leaf_indicator
+    content_tag(:span, tabindex: 0, class: 'tree-menu--leaf-indicator') do
+      content_tag(:span,
+                  'Hierarchy leaf',
```
Comment (oliverguenther) [inline app/helpers/application_helper.rb]: Missing I18n

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -573,6 +570,59 @@ def permitted_params
...
+                     project_wiki_path(page.project, page),
+                     title: hierarchy_item_title(options, page),
+                     class: 'tree-menu--title ellipsis') do
+      end
+    end
+  end
+
+  def hierarchy_item_title(options, page)
+    if options[:timestamp] && page.updated_on
+      l(:label_updated_time, distance_of_time_in_words(Time.now, page.updated_on))
```
Comment (oliverguenther) [inline app/helpers/application_helper.rb]: Prefer standard I18n's `t(:...)` over redmine's magic `l(:....)`.

File: config/initializers/menus.rb [hunk 10]
```diff
@@ -287,6 +293,24 @@
...
+  #           param: :project_id,
+  #           caption: :label_wiki_start,
+  #           parent: :wiki
+  #
+  # menu.push :wiki_menu_partial,
+  #           { controller: '/wiki', action: 'show' },
+  #           param: :project_id,
+  #           parent: :wiki,
+  #           partial: 'wiki/menu_pages_tree'
+
```
Comment (oliverguenther) [inline config/initializers/menus.rb]: Please remove the commented code and replace with a single comment a la `# Wiki menu items are added by WikiMenuItemHelper`.

File: frontend/app/angular4-modules.ts [hunk 11]
```diff
@@ -542,16 +544,23 @@ import {WorkPackageRelationsCreateComponent} from "core-components/wp-relations/
...
   }
 
   // noinspection JSUnusedGlobalSymbols
-  ngDoBootstrap() {
+  ngDoBootstrap(appRef:ApplicationRef) {
     // Already done in openproject-app.ts
     // this.upgrade.bootstrap(document.body, ['openproject'], {strictDi: false});
+
+    if (document.getElementsByTagName('main-menu-resizer').length > 0) {
+      appRef.bootstrap(MainMenuResizerDirective);
```
Comment (oliverguenther) [inline frontend/app/angular4-modules.ts:562]: Why didn't you use `downgradeComponent` in this case?

File: frontend/app/components/resizer/main-menu-resizer.component.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+  private resizeEvent:string;
+  private localStorageKey:string;
+
+  private resizingElement:HTMLElement;
+  private elementWidth:number;
+  private oldPosition:number;
+  private mouseMoveHandler:any;
+  private element:HTMLElement;
+  private htmlNode:HTMLElement;
+  private toggleTitle:string;
```
Comment (oliverguenther) [inline frontend/app/components/resizer/main-menu-resizer.component.ts]: Properties in the view should be public, not private. This is irrelevant during runtime since these modifiers are only for type-checking. But they will raise errors when AOT is enabled.

File: frontend/app/components/resizer/main-menu-resizer.component.ts [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+  constructor(private elementRef:ElementRef) {
+  }
+
+  ngOnInit() {
+    this.elementClass    = "main-menu";
+    this.resizeEvent     = "main-menu-resize";
+    this.localStorageKey = "openProject-mainMenuWidth";
+    this.toggleTitle     = I18n.t('js.show_hide_project_menu');
+
+    this.htmlNode = <HTMLElement>document.getElementsByTagName('html')[0];
```
Comment (oliverguenther) [inline frontend/app/components/resizer/main-menu-resizer.component.ts:70]: Prefer `document.documentElement`

File: frontend/app/components/resizer/main-menu-resizer.component.ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    // Get element
+    this.resizingElement = <HTMLElement>document.getElementsByClassName(this.elementClass)[0];
```
Comment (oliverguenther) [inline frontend/app/components/resizer/main-menu-resizer.component.ts:73]: Prefer `document.getElementsByClassName(this.elementClass)[0] as HTMLElement`

File: frontend/app/components/resizer/main-menu-resizer.component.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+      this.elementWidth = this.resizingElement.clientWidth;
+      this.moving = true;
+
+      // Necessary to encapsulate this to be able to remove the event listener later
+      this.mouseMoveHandler = this.resizeElement.bind(this, this.resizingElement);
+
+      // Change cursor icon
+      // This is handled via JS to ensure
+      // that the cursor stays the same even when the mouse leaves the actual resizer.
+      document.getElementsByTagName("body")[0].setAttribute('style',
```
Comment (oliverguenther) [inline frontend/app/components/resizer/main-menu-resizer.component.ts]: Prefer `document.body.setAttribute...`

File: frontend/app/layout/controllers/main-menu-controller.js [hunk 16]
```diff
@@ -27,36 +27,23 @@
 //++
 
 module.exports = function($rootScope, $window) {
-  var menuList = null;
+  var htmlNode = document.getElementsByTagName('html')[0];
```
Comment (oliverguenther) [inline frontend/app/layout/controllers/main-menu-controller.js]: Prefer `document.documentElement`

File: lib/redmine/menu_manager/menu_helper.rb [hunk 17]
```diff
@@ -77,10 +77,20 @@ def display_main_menu?(project)
...
+    menu_items = menu_items_for(menu, project) do |node|
       links << render_menu_node(node, project)
     end
-    links.empty? ? nil : content_tag('ul', links.join("\n").html_safe, class: 'menu_root')
+    if menu == :project_menu
+      selected = any_item_selected?(select_leafs(menu_items))
+      classes << (selected ? 'open' : 'closed')
+    end
+
+    links.empty? ? nil : content_tag('ul', links.join("\n").html_safe, class: 'menu_root ' + classes)
```
Comment (oliverguenther) [inline lib/redmine/menu_manager/menu_helper.rb]: If possible, use `safe_join(links, "\n")` instead to avoid forcing them to be html_safe

File: lib/redmine/menu_manager/menu_helper.rb [hunk 18]
```diff
@@ -77,10 +77,20 @@ def display_main_menu?(project)
 
   def render_menu(menu, project = nil)
     links = []
-    menu_items_for(menu, project) do |node|
+    classes = ''
+    menu_items = menu_items_for(menu, project) do |node|
       links << render_menu_node(node, project)
     end
-    links.empty? ? nil : content_tag('ul', links.join("\n").html_safe, class: 'menu_root')
```
Comment (oliverguenther) [inline lib/redmine/menu_manager/menu_helper.rb]: If possible, use `safe_join(links, "\n")` instead to avoid forcing them to be html_safe

File: lib/redmine/menu_manager/menu_helper.rb [hunk 19]
```diff
@@ -131,7 +141,7 @@ def render_drop_down_label_node(label, selected, options = {})
   end
 
   def any_item_selected?(items)
-    items.any? { |item| item.name == current_menu_item }
+    items.any? { |item| item.name == current_menu_item || item.name == "entry-item-#{current_menu_item}".to_sym }
```
Comment (oliverguenther) [inline lib/redmine/menu_manager/menu_helper.rb]: Use the `entry_page_selected?(item)` helper from below in this module.

File: lib/redmine/menu_manager/wiki_menu_helper.rb [hunk 20]
```diff
@@ -67,4 +73,32 @@ def push_wiki_menu_subitem(menu, main_item, child)
...
+    menu.push :wiki_menu_partial,
+              { controller: '/wiki', action: 'show' },
+              param: :project_id,
+              parent: main_item.menu_identifier,
+              partial: 'wiki/menu_pages_tree',
+              last: true
+  end
+
+  def push_wiki_entry_page_item(main_item, menu)
+    menu.push "entry-item-#{main_item.menu_identifier}".to_sym,
```
Comment (oliverguenther) [inline lib/redmine/menu_manager/wiki_menu_helper.rb]: You may want to extract `"entry-item-#{item.menu_identifier}".to_sym` into a separate helper method since it occurs in many places

File: spec_legacy/fixtures/menu_items.yml [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+---
+menu_items_001:
```
Comment (oliverguenther) [inline spec_legacy/fixtures/menu_items.yml:30]: Is this file really needed? We shouldn't create new fixtures

**Review by oliverguenther — CHANGES_REQUESTED** [review]
There are a few minor code remarks I'd like to have addressed. Overall, this is very solid. Can be merged afterwards.

We should extract the accessibility issues into a separate PR.

File: app/helpers/application_helper.rb (hunk 9 above)
Comment (wielinde) [reply app/helpers/application_helper.rb]: @oliverguenther Yeah, but I18n's @t(:...)@ didn't work here. Might be as this is a date/time localisation and thus a different story than translation?

Comment (oliverguenther) [reply app/helpers/application_helper.rb]: It will work if you use `t(:label_updated_time, value: distance_of_time...)`.

File: app/assets/javascripts/tree_menu.js (hunk 6 above)
Comment (wielinde) [reply app/assets/javascripts/tree_menu.js]: Great!!

File: frontend/app/angular4-modules.ts (hunk 11 above)
Comment (wielinde) [reply frontend/app/angular4-modules.ts:562]: I understand that too little. Why should we downgrade to AngularJS the main-menu-resizer is an Angular entry component and does not run within AngularJS? Probably I don't understand it enough.

File: frontend/app/components/resizer/main-menu-resizer.component.ts (hunk 12 above)
Comment (wielinde) [reply frontend/app/components/resizer/main-menu-resizer.component.ts]: Ok, you mean those properties that appear in curly braces in the template, thus only `toggleTitle`, right?

File: spec_legacy/fixtures/menu_items.yml (hunk 21 above)
Comment (wielinde) [reply spec_legacy/fixtures/menu_items.yml:30]: To my defense: I felt bad to do that. However, I do not want to migrate the cukes at this point of time.
